macq/extract/locm.py

```diff
@@ -0,0 +1,802 @@
+""".. 'include':: ../../docs/templates/extract/locm.md"""
+
+from collections import defaultdict
+from dataclasses import asdict, dataclass
+from pprint import pprint
+from typing import Dict, List, NamedTuple, Optional, Set, Tuple, Union
+from warnings import warn
+
+from macq.trace.action import Action
+from macq.trace.fluent import PlanningObject
+
+from ..observation import ActionObservation, Observation, ObservedTraceList
+from . import LearnedLiftedAction, Model
+from .exceptions import IncompatibleObservationToken
+from .learned_fluent import LearnedLiftedFluent
+from .model import Model
+
+
+@dataclass
+class AP:
+    """Action.Position (of object parameter). Position is 1-indexed."""
+
+    action: Action
+    pos: int  # NOTE: 1-indexed
+    sort: int
+
+    def __repr__(self) -> str:
+        return f"{self.action.name}.{self.pos}"
+
+    def __hash__(self):
+        return hash(self.action.name + str(self.pos))
+
+    def __eq__(self, other):
+        return hash(self) == hash(other)
+
+
+class StatePointers(NamedTuple):
+    start: int
+    end: int
+
+    def __repr__(self) -> str:
+        return f"({self.start} -> {self.end})"
+
+
+Sorts = Dict[str, int]  # {obj_name: sort}
```

Comment:
Looks like a lot of top-level globals...what happens if we want to run locm on a few different data sets without creating a new python env fresh?

Reply:
globals are all just type references. allows typing things as `APStatePointers' for example instead of littering the code with `Dict[int, Dict[AP, Tuple[int, int]]]` which is way more confusing when just glancing at call signatures.

running on multiple datasets will work fine

Reply:
Got it, thanks!